---
title: "Phyloseq PreProcessing"
author: "Calla Bush St George"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/02_PreProcessing/") # send any figure output to this folder 
```


# Load libraries 
```{r load-libraries}
#devtools::install_github("joey711/phyloseq")


pacman::p_load(devtools, phyloseq, tidyverse, dada2, readr, install = FALSE)
```

# Goals

Here, we will process the data into a phyloseq object. 

- ASV table 
- Taxonomy Table 
- Track Reads (metadata)

Then, we will remove the following: 

1. Remove Chloroplasts
2. Remove Mitochondria. 
3. Removing ASVs from negative controls and also negative control samples.
4. Remove samples without "enough" reads. 

Finally, write data file of phyloseq output.


# Load Data 

## ASV Table
```{r load-asv-table}
# First, load asv table
asv_tab <- readr::read_tsv("data/01_DADA2/04_end_files/ASV_counts_epulos.tsv")

# Inspect asv_tab
head(asv_tab)[,1:5]

# Fix names 
sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"),
                        FUN = function(x)paste(x[1],sep="_"))
head(samples_fixed)



# re-write the ASV count file to fix names 
colnames(asv_tab) <- samples_fixed
str(asv_tab)
```

##  Taxonomy Table
```{r load-tax-table}
tax_df <- read.table("data/01_DADA2/04_end_files/ASV_taxonomy_epulos.tsv",
                     sep = "\t", skip = 1)

head(tax_df)

# fix column names 
colnames(tax_df) <- 
  c("asv_names", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus",
    "Species", "ASV", "ASVseq")

head(tax_df)

# Taxonomy Table Matrix
tax_mat <- 
  tax_df %>%
  tibble::column_to_rownames(., var = "asv_names") %>%
  as.matrix()


asv_mat <- 
  asv_tab %>%
  tibble::column_to_rownames(., var = "names") %>%
  as.matrix()
```


##  Track Reads Data
```{r load-metadata}
# Load in metadata

metadata_df <- read.csv("data/metadata.csv")
dim(metadata_df)
colnames(metadata_df)

# Update row.names to be sample names for metadata
## Before 
row.names(metadata_df)
# Rewrite 
row.names(metadata_df) <- metadata_df$names
# Check afterwards that it worked 
row.names(metadata_df)
# intuition check
head(metadata_df)

```


# Handoff to phyloseq
```{r phyloseq-handoff}
# double check it's all good 
dim(asv_mat)
dim(tax_mat)

# Intuition check 
stopifnot(row.names(asv_mat) == row.names(tax_mat))

# Construct the phyloseq object 
raw_physeq <- phyloseq(otu_table(asv_mat, taxa_are_rows = TRUE),
                       sample_data(metadata_df),
                       tax_table(tax_mat))
raw_physeq

# Save this raw phyloseq object 
# Make new folder in data/ "02_Preprocessing"
save(raw_physeq, file = "data/06_PreProcessing_epulos/raw_physeq.RData")
```

# Clean up the data
Remove: 

1. Chloroplasts
2. mitochondria  

```{r rm-mitos-chloros}
# Remind myself of tax table 
#View(tax_mat)

# Make new physeq without chloroplasts
noChloros_physeq <- 
  raw_physeq %>% 
  # rm chloroplasts
  subset_taxa(Order != "Chloroplast" | is.na(Order))
  
# How many taxa were chloroplasts? 
num_chloro_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloro_ASVs

# Intuition chek 
#noChloros_physeq %>%
#  tax_table() %>%
#  data.frame() %>%
#  View()

# remove mitochondria 
noChlorosMitos_physeq <- 
  noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

# How many mitochondrial ASVs? 
num_mito_ASVs <- ntaxa(noChloros_physeq) - ntaxa(noChlorosMitos_physeq)
num_mito_ASVs

noChlorosMitos_physeq

# How many total asvs were removed from chloros and mitos 
ntaxa(raw_physeq) - ntaxa(noChlorosMitos_physeq)
# proportion of asvs kept? 
ntaxa(noChlorosMitos_physeq)/ntaxa(raw_physeq)
```

# Evaulate and remove the control samples 

Take a look at the negative controls and then make a decision about whether or not to remove the ASVs that we found in our controls 

1. Negative controls 
2. ASVs found within the negative controls and their distribution in the samples. 
3. Evaluate the mock community 


```{r seq-depth}
# The current data object
noChlorosMitos_physeq

# What is the library size/sequencing depth for each sample? 
seqSums_df <- 
  noChlorosMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_df, by = "names") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)
head(seqSums_df)

# Show the depth of samples 
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

# plot it! 
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs,
             fill = gut_section)) + 
  geom_bar(stat = "identity") 

# Density plot 
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = gut_section)) +
  geom_density(alpha = 0.5)

```

# Save Preprocessed Phyloseq Object
```{r save-physeq}

raw_preprocessed_physeq <- noChlorosMitos_physeq

save(raw_preprocessed_physeq,
     file = "data/06_PreProcessing_epulos/raw_preprocessed_physeq.RData")
```

# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```

